File: example/android/StarPrinterLib/src/main/java/com/getzuza/starprinterlib/Receipt.java
package com.getzuza.starprinterlib;

import android.content.Context;
import android.graphics.Bitmap;
import android.graphics.Typeface;
import android.media.Image;

import com.starmicronics.starioextension.ICommandBuilder;

import java.nio.charset.Charset;

public  class Receipt {

    protected PaperSizes _size;

    protected Languages _languages;

    protected ICommandBuilder _builder;

    protected Context _context;

    private boolean _closed;

    public static Receipt createReceiptFromText(Context context, boolean text, Languages language, int paperSize) {

        PaperSizes size = PaperSizes.None;
        Receipt receipt;
        if (text) {
            switch (paperSize) {
                case 2:
                    size = PaperSizes.PaperSizeTwoInchText;
                    break;
                case 3:
                    size = PaperSizes.PaperSizeThreeInchText;
                    break;
                case 4:
                    size = PaperSizes.PaperSizeFourInchText;
                    break;
                default:
                    size = PaperSizes.PaperSizeThreeInchText;
                    break;
            }
            receipt = new TextReceipt(context, language, size);
        } else {
            switch (paperSize) {
                case 2:
                    size = PaperSizes.PaperSizeTwoInch;
                    break;
                case 3:
                    size = PaperSizes.PaperSizeThreeInch;
                    break;
                case 4:
                    size = PaperSizes.PaperSizeFourInch;
                    break;
                default:
                    size = PaperSizes.PaperSizeThreeInch;
                    break;
            }
            receipt = new GraphicReceipt(context, language, size);
        }
        receipt._builder.beginDocument();
        return receipt;
    }

    public void setFont(Typeface font) {

    }

    public void setFontName(String name) {

    }

    public void setFontSize(float size) {
    }

    public void addAlignLeft() {
    }

    public void addAlignRight() {
    }

    public void addAlignCenter() {
    }

    public void addText(String text) {
    }

    public void addDoubleText(String text) {
    }

    public void addBoldText(String text) {
    }

    public void addUnderlinedText(String text) {
    }

    public void addInverseText(String text) {
    }

    public void addLine() {
    }

    public void addBarcode(String text, int height) {
        text = String.format("{B%s", text);
        _builder.appendBarcodeWithAlignment(text.getBytes(),
                ICommandBuilder.BarcodeSymbology.Code128,
                ICommandBuilder.BarcodeWidth.ExtMode1,
                height, true, ICommandBuilder.AlignmentPosition.Center);
    }

    protected int paperWidth() {
        switch (_size) {
            default:
            case PaperSizeDotImpactThreeInch:
            case PaperSizeThreeInchText:
            case PaperSizeThreeInch:
            case PaperSizeEscPosThreeInch:
                return PaperSizes.PaperSizeThreeInch.getSize();

            case PaperSizeFourInch:
            case PaperSizeFourInchText:
                return PaperSizes.PaperSizeFourInch.getSize();

            case PaperSizeTwoInch:
            case PaperSizeTwoInchText:
                return PaperSizes.PaperSizeTwoInch.getSize();
        }
    }

    public void addImage(Bitmap image, int width) {
        if (image != null) {
            int w = paperWidth();
            if (width <= 0 || width > w) {
                width = w;
            }
            _builder.appendBitmapWithAlignment(image, false, width, true, ICommandBuilder.AlignmentPosition.Center);
        }
    }

    public void addQrCode(String data) {
        if (data != null) {
            byte[] bytes = data.getBytes();
            _builder.appendQrCodeWithAlignment(bytes, ICommandBuilder.QrCodeModel.No2, ICommandBuilder.QrCodeLevel.M,8, ICommandBuilder.AlignmentPosition.Center);
            //_builder.appendBitmapWithAlignment(image, false, width, true, ICommandBuilder.AlignmentPosition.Center);
        }
    }

    public void openCashDrawer() {
        _builder.endDocument();
        _builder.appendPeripheral(ICommandBuilder.PeripheralChannel.No1);
        _builder.appendPeripheral(ICommandBuilder.PeripheralChannel.No2);
        _builder.beginDocument();
    }

    public void cutPaper() {
        _builder.appendCutPaper(ICommandBuilder.CutPaperAction.PartialCutWithFeed);
    }

    public void setBlackColor() {
    }

    public void setRedColor() {
    }

    public void closeReceipt() {
        if (!_closed) {
            _builder.endDocument();
            _closed = true;
        }
    }

    public void setCommand(String command) {
        byte[] data;
        try {
            data = command.getBytes("UTF-8");
        } catch (Exception e) {
            data = new byte[0];
        }
        _builder.append((data));
    }

}

File: ios/Classes/Receipt.m
//
//  Receipt.m
//  StarPrinter
//
//  Created by Shailesh Lolam on 11/8/17.
//  Copyright © 2017 Shailesh Lolam. All rights reserved.
//

#import <Foundation/Foundation.h>
#import "Receipt.h"
#import "TextReceipt.h"
#import "GraphicReceipt.h"

@implementation Receipt
@synthesize builder;
@synthesize language;
@synthesize paperSize;
bool isClosed = false;

+(Receipt *) createReceiptFromText: (bool) isText Language: (Languages) language AndPaperSize: (int) paperSize;{
    Receipt *receipt = nil;
    PaperSizes size;
    NSLog(@"Creating receipt for text=%d", isText);
    if(isText){
        switch (paperSize) {
            case 2:
                size = PaperSizeTwoInchText;
                break;
            case 3:
                size = PaperSizeThreeInchText;
                break;
            case 4:
                size = PaperSizeFourInchText;
                break;
            default:
                size = PaperSizeThreeInchText;
                break;
        }
        receipt = [[TextReceipt alloc] initWithLanguage:language AndSize:size];
    } else{
        switch (paperSize) {
            case 2:
                size = PaperSizeTwoInch;
                break;
            case 3:
                size = PaperSizeThreeInch;
                break;
            case 4:
                size = PaperSizeFourInch;
                break;
            default:
                size = PaperSizeThreeInch;
                break;
        }
        receipt = [[GraphicReceipt alloc] initWithLanguage:language AndSize:size];
    }

    [receipt.builder beginDocument];
    return receipt;
}

-(void) setFont: (UIFont *) font{
    
}

-(void) setFontName: (NSString *) name{
    
}

-(void) setFontSize: (CGFloat) size{
    
}

-(void) addAlignLeft{
    
}
-(void) addAlignCenter{
    
}
-(void) addAlignRight{
    
}
-(void) addText: (NSString *) string{
    
}

-(void) addDoubleText: (NSString *) string{
    
}

-(void) addBoldText: (NSString *) string{
    
}

-(void) addUnderlinedText: (NSString *) string{
    
}

-(void) addInverseText: (NSString *) string{
    
}

-(void) addBarcode: (NSString *) string  height: (int) height{
    string = [NSString stringWithFormat:@"{B%@", string];
    [self.builder appendBarcodeDataWithAlignment:[string dataUsingEncoding:NSASCIIStringEncoding] symbology:SCBBarcodeSymbologyCode128 width:SCBBarcodeWidthMode2 height:height hri:YES position:SCBAlignmentPositionCenter];
}

-(int) paperWidth {
    switch (paperSize) {
        default:
        case PaperSizeDotImpactThreeInch:
        case PaperSizeThreeInchText:
        case PaperSizeThreeInch:
        case PaperSizeEscPosThreeInch:
            return PaperSizeThreeInch;

        case PaperSizeFourInch:
        case PaperSizeFourInchText:
            return PaperSizeFourInch;
            
        case PaperSizeTwoInch:
        case PaperSizeTwoInchText:
            return PaperSizeTwoInch;
    }
}

-(void) addImage: (UIImage *) source width:(int)width{
    //[self drawPrevious];
    if(source){
        int w = [self paperWidth];
        if(width <= 0 || width > w) {
            width = w;
        }
        [self.builder appendBitmapWithAlignment:source diffusion:NO width:width bothScale:YES position:SCBAlignmentPositionCenter];
    }
}

-(void) addQrCode: (NSString *) data{
    if(data){

        NSData* raw = [data dataUsingEncoding:1];
        
        [self.builder appendQrCodeDataWithAlignment:raw model:SCBQrCodeModelNo2 level:SCBQrCodeLevelM cell:8 position:SCBAlignmentPositionCenter];
    }
}


-(void) openCashDrawer: (int) drawer{
    [self.builder endDocument];
    if (drawer > 1) {
        [self.builder appendPeripheral:SCBPeripheralChannelNo2];
    } else {
        [self.builder appendPeripheral:SCBPeripheralChannelNo1];
    }
    [self.builder beginDocument];
}

-(void) cutPaper{
    [self.builder appendCutPaper:SCBCutPaperActionPartialCutWithFeed];
}

-(void) setBlackColor{
}

-(void) setRedColor{
}

-(void) closeReceipt{
    if(!isClosed) {
        [[self builder] endDocument];
        isClosed = true;
    }
}

-(void) setCommand: (NSString *) command {
        const void *cmd = (const void *)[command UTF8String];
        [self.builder appendBytes:cmd length:command.length];
}

-(void) addLine{
    [self addText: @"--------------------\n"];
}

+(void) beginGraphicContext: (CGSize) size {
    if ([UIScreen.mainScreen respondsToSelector:@selector(scale)]) {
        if (UIScreen.mainScreen.scale == 2.0) { //Retina screen
            UIGraphicsBeginImageContextWithOptions(size, NO, 1.0);
        } else { //Non-retina
            UIGraphicsBeginImageContext(size);
        }
    } else {
        UIGraphicsBeginImageContext(size);
    }
}

+(UIImage *)imageWithString:(NSString *)string Width:(CGFloat)width Attributes:(NSDictionary*) attributes{
    
    //string = [string stringByTrimmingCharactersInSet:[NSCharacterSet whitespaceCharacterSet]];
    
    CGSize size = [string boundingRectWithSize:CGSizeMake(width, 1000)
                                       options:NSStringDrawingUsesLineFragmentOrigin | NSStringDrawingTruncatesLastVisibleLine
                                    attributes:attributes
                                       context:nil].size;
    //Reset the width to full
    size.width = width;
    
    [self beginGraphicContext:size];
    
    CGContextRef context = UIGraphicsGetCurrentContext();
    
//    CGContextSetFillColorWithColor(context, [(UIColor*)[attributes objectForKey:NSForegroundColorAttributeName] CGColor]);
//
//    CGContextSetStrokeColorWithColor(context, [(UIColor*)[attributes objectForKey:NSBackgroundColorAttributeName] CGColor]);
    
    [(UIColor*)[attributes objectForKey:NSBackgroundColorAttributeName] set];

    //[(UIColor*)[attributes objectForKey:NSForegroundColorAttributeName] setStroke];

    CGRect rect = CGRectMake(0, 0, size.width + 1, size.height + 1);
    
    CGContextFillRect(context, rect);
    
    [string drawInRect:rect withAttributes:attributes];
    
    UIImage *imageToPrint = UIGraphicsGetImageFromCurrentImageContext();
    
    UIGraphicsEndImageContext();
    
    return imageToPrint;
}

+(UIImage *)imageWithString:(NSAttributedString *)string Width:(CGFloat)width {
    
    if (!string || string.length == 0){
        return nil;
    }
    
    CGSize size = [string boundingRectWithSize:CGSizeMake(width, CGFLOAT_MAX)
                                       options:NSStringDrawingUsesLineFragmentOrigin | NSStringDrawingTruncatesLastVisibleLine
                                    //attributes:attributes
                                       context:nil].size;

    //Reset the width to full
    size.width = width;
    
    [self beginGraphicContext:size];
    
    CGContextRef context = UIGraphicsGetCurrentContext();
    
    [[UIColor whiteColor] set];

    CGRect rect = CGRectMake(0, 0, size.width + 1, size.height + 1);
    
    CGContextFillRect(context, rect);
    
    [string drawInRect:rect];
    
    UIImage *imageToPrint = UIGraphicsGetImageFromCurrentImageContext();
    
    UIGraphicsEndImageContext();
    
    return imageToPrint;
}
@end
